File: protein-translation/protein_translation.py
TRANSLATE = {
    'Methionine': ['AUG'],
    'Phenylalanine': ['UUU', 'UUC'],
    'Leucine': ['UUA', 'UUG'],
    'Serine': ['UCU', 'UCC', 'UCA', 'UCG'],
    'Tyrosine': ['UAU', 'UAC'],
    'Cysteine': ['UGU', 'UGC'],
    'Tryptophan': ['UGG'],
    'STOP': ['UAA', 'UAG', 'UGA'],
}


def proteins(strand):
    """Translates RNA sequences into proteins."""
    polypep = []
    codons = [strand[x:x+3] for x in range(0, len(strand), 3)]
    for codon in codons:
        if codon in TRANSLATE['STOP']:
            break
        for k, v in TRANSLATE.items():
            if codon in v:
                polypep.append(k)
    return polypep

File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
    {
        return Chunks(strand)
            .Select(Translate)
            .TakeWhile(protein => protein != "STOP")
            .ToArray();
    }

    private static IEnumerable<string> Chunks(string strand)
    {
        if (strand.Length is var length && length % 3 != 0)
        {
            throw new ArgumentException($"Invalid strand length: {length}.");
        }

        for (int i = 0; i < length; i += 3)
        {
            yield return strand.Substring(i, 3);
        }
    }

    private static string Translate(string codon)
    {
        return codon switch
        {
            "AUG" => "Methionine",
            "UUC" or "UUU" => "Phenylalanine",
            "UUA" or "UUG" => "Leucine",
            "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
            "UAU" or "UAC" => "Tyrosine",
            "UGU" or "UGC" => "Cysteine",
            "UGG" => "Tryptophan",
            "UAA" or "UAG" or "UGA" => "STOP",
            _ => throw new ArgumentException($"Invalid codon: {codon}."),
        };
    }
}
